Add a version chooser to DDoc pages

wilzbach opened the pull request:
From the [NG](http://forum.dlang.org/post/niuatqnkrygvunacwjti@forum.dlang.org):

> IMO there should be a link under the "switch to pre-release." link in the current docs.
> Maybe replace it with a dropdown to select the version from 2.066 to pre-release?

CC @JackStouffer @CyberShadow 


![image](https://user-images.githubusercontent.com/4370550/27890522-7750bf26-61f3-11e7-936c-997f7abc09fc.png)

![image](https://user-images.githubusercontent.com/4370550/27890526-7e9a85be-61f3-11e7-90ec-12eefaf43410.png)

![image](https://user-images.githubusercontent.com/4370550/27890567-c14d5044-61f3-11e7-9da7-84436d490d60.png)

It's not perfect yet (e.g. I need to detect the current version automatically and add the same for the Ddox build), but I wanted to get your opinion on this before I invest more work.

CyberShadow commented:
The dropdown will be absent in your old builds, right? So it will only be present on dlang.org?

wilzbach commented:
> The dropdown will be absent in your old builds, right? So it will only be present on dlang.org?

For now, yes, though we could always patch the archived files.

CyberShadow commented:
Needs a bit of margin to the left:
 ![](https://dump.thecybershadow.net/0d0c2b3ba566135111f065e41efb9631/00%3A22%3A00-upload.png)

Why place it inside the content and not to the right of the module title?
![](https://dump.thecybershadow.net/d8878e8299d69d4f2a928a2615189052/00%3A23%3A21-upload.png)

CyberShadow commented:
@andralex You've been asking for this since forever, what do you think of this implementation?

CyberShadow commented:
Being able to go to the *corresponding* page between DDoc/DDox is awesome! But it seems a little buggy: for http://dtest.dlang.io/artifact/website-191526224e06ec8685ef43831cbe43983d14ec25-b0341c56fbd877b803c357bf95312672/web/library/core/atomic/atomic_fence.html the dropdown has "stable" selected by default, and if I select ddox again, it goes to a 404.

andralex commented:
Hook, line, and sinker! But there should be no ddox vs ddoc distinction. Instead, each should have a version chooser. No? Thanks @wilzbach!!

wilzbach commented:
> Why place it inside the content and not to the right of the module title?

I did - just get a proper browser ;-)

> Being able to go to the corresponding page between DDoc/DDox is awesome! But it seems a little buggy: 

I never intended it to work on DDox - cleaned up the code and made it Ddoc/Ddox agnostic :)

> Hook, line, and sinker! But there should be no ddox vs ddoc distinction. Instead, each should have a version chooser. No? 

That's actually one of the feature that I have missed myself for a long line. It's extremely handy and raises the awareness of Ddox. OTOH even if we switch to Ddox as default, which we hopefully will do within the next months, people still might want to switch back to Ddoc.
However, I renamed it from "(ddox)" to "(beta)", s.t. people aren't that confused.

CyberShadow commented:
> I did - just get a proper browser ;-)

What counts for a proper browser, then?

![](https://dump.thecybershadow.net/d4dc72ec2285931ad0b55c3182dd0efd/01%3A38%3A30-upload.png)

wilzbach commented:
One that behaves as my local ones - or in other words "one that doesn't have bugs" :smiling_imp: 

Chrom(ium):

![image](https://user-images.githubusercontent.com/4370550/27939708-e9b27bd0-62c6-11e7-9330-b8ced0128a5b.png)

Firefox:

![image](https://user-images.githubusercontent.com/4370550/27939695-dd399cbc-62c6-11e7-9a78-6ae4a18e4a5c.png)

Maybe you have more luck with the latest version.

CyberShadow commented on js/listanchors.js:
```diff
@@ -88,6 +88,65 @@ function listanchors()
+      var options = versions.map(function(e, i){
+        return "<option value='" + i + "'" + (e.selected ? "selected" : "") + ">" + e.name + "</option>";
+      });
+      $("h1").after("<div class='version-changer-container fa-select'><select id='version-changer'>" + options.join("") + "</select></div>");
```
I'm pretty sure you're supposed to put the element you're floating *before* the element it is supposed to "overlap", not `after`.

Why does this work for you?

wilzbach replied:
> I'm pretty sure you're supposed to put the element you're floating before the element it is supposed to "overlap", 

I tried `before` as well, but discarded the approach due as floating didn't look nice for mobile devices. Though of course, now I already added viewport-specific CSS.
For reference, currently it looks like this:

![image](https://user-images.githubusercontent.com/4370550/27942899-23b562c0-62db-11e7-8db8-8e6d1e8ef820.png)

With `before` it could only look like this:

![image](https://user-images.githubusercontent.com/4370550/27942893-1b13dd22-62db-11e7-9ff3-54dd2c120539.png)

Though of course, we could simple hide the select box on mobile devices.

> Why does this work for you?

```css
.version-changer-container {
    margin-top: -4em;
}
```

The interesting question is why it doesn't work for ;-)
A couple of external tools show the same results:

https://tools.pingdom.com/#!/bpOIFJ/http://dtest.dlang.io/artifact/website-b29c592343b3eca4d1a2bb7774123c738e67b788-7991f163dc4768833cc161ca4595703d/web/phobos-prerelease/core_bitop.html
https://www.webpagetest.org/result/170707_0E_8F6/

Google even allows to preview the mobile viewport (this one has special CSS - no floating):

https://developers.google.com/speed/pagespeed/insights/?url=http%3A%2F%2Fdtest.dlang.io%2Fartifact%2Fwebsite-b29c592343b3eca4d1a2bb7774123c738e67b788-7991f163dc4768833cc161ca4595703d%2Fweb%2Fphobos-prerelease%2Fcore_bitop.html

CyberShadow commented:
Oh, you fixed it!

Old (broken): http://dtest.dlang.io/artifact/website-191526224e06ec8685ef43831cbe43983d14ec25-b0341c56fbd877b803c357bf95312672/web/library/core/atomic/atomic_fence.html
New (works): http://dtest.dlang.io/artifact/website-b29c592343b3eca4d1a2bb7774123c738e67b788-7991f163dc4768833cc161ca4595703d/web/library-prerelease/core/atomic/atomic_fence.html

CyberShadow commented:
Well that was confusing. Sneakily fixing it and then blaming my browser. How much is Google paying you, eh ;)

http://dtest.dlang.io/artifact/website-b29c592343b3eca4d1a2bb7774123c738e67b788-7991f163dc4768833cc161ca4595703d/web/library/std/algorithm/comparison/among.html doesn't recognize it's on a DDox page. Selecting the "beta" option takes you to the module's page.

I noticed you changed "DDox" to "master (beta)". I was going to say something about the DDox name being an implementation detail, but I figured that space is tight anyway and it might not hurt to let it build up a brand name... However "master (beta)" may be confusing because it's not clear that it's a beta of the documentation generator, and not of D itself.

wilzbach commented:
> Well that was confusing. Sneakily fixing it and then blaming my browser. How much is Google paying you, eh ;)

Not enough!
I [did state this](https://github.com/dlang/dlang.org/pull/1801#issuecomment-313567140):

>> Maybe you have more luck with the latest version.

> http://dtest.dlang.io/artifact/website-b29c592343b3eca4d1a2bb7774123c738e67b788-7991f163dc4768833cc161ca4595703d/web/library/std/algorithm/comparison/among.html doesn't recognize it's on a DDox page. Selecting the "beta" option takes you to the module's page.

Hmm symbol pages are currently not supported as I wanted to go for the the simplest, working solution for now. Detecting the current symbol shouldn't be that hard - it's essentially a one-line

```js
var symbol = (location.href.split(ddoxModuleURL.replace(".html", ""))[1] || "").replace("/", "").split(".html")[0];
```

But looking at the output it seems that more CSS tweaking would be required:

![image](https://user-images.githubusercontent.com/4370550/27943591-da2f9ffc-62e0-11e7-97cc-a305ef5f24db.png)

Hence, I would leave showing the select box on a symbol page for a future improvement and simply hide the select box if the detection logic failed.

> I noticed you changed "DDox" to "master (beta)". I was going to say something about the DDox name being an implementation detail, but I figured that space is tight anyway and it might not hurt to let it build up a brand name... However "master (beta)" may be confusing because it's not clear that it's a beta of the documentation generator, and not of D itself.

Fair point, [I only changed it](https://github.com/dlang/dlang.org/pull/1801#issuecomment-313563540) because Andrei didn't like "Ddox" at all. I wasn't a huge friend of "beta" -> changing it back then.

CyberShadow commented:
> Maybe you have more luck with the latest version.

Oops, I thought you meant the latest version of the web browser! I'm using Firefox Developer.

> Fair point, I only changed it because Andrei didn't like "Ddox" at all. I wasn't a huge friend of "beta" -> changing it back then.

Yeah, I can't think of better solutions right now.

I guess "beta" + a banner on DDox pages explaining why it's in beta and where to report issues would work, but it seems like a relatively minor issue with no obviously correct solution anyway.

wilzbach commented:
> I guess "beta" + a banner on DDox pages explaining why it's in beta and where to report issues would work, but it seems like a relatively minor issue with no obviously correct solution anyway.

Yup, especially because we haven't shown such a banner before, so I guess I leave finding a "correct solution" up to another PR.

So are we good to go here @andralex?

andralex commented:
That the space is tight doesn't make "ddox" any better. It's an internal name that tells nothing to the casual observer. I'll approve this but definitely we need to revisit the matter in the future.

CyberShadow commented:
@andralex I think we are all in agreement, just lacking a better solution.
